### PR TITLE
Extensible name section

### PR DESCRIPTION
**Pull request by pipcet**
First stab at addressing https://github.com/WebAssembly/binaryen/issues/914.

No locals yet, but we weren't emitting those before.

Are there any tests I should be running?

### CODE REVIEW

**Comment by kripken** [comment]
Thanks! I didn't look closely at the code yet, but to verify this works, the binaryen tests can be run with `./check.py  --no-test-waterfall`, and emscripten has specific tests for this, `./tests/runner.py other.test_binaryen_debuginfo` and `./tests/runner.py  binaryen*.test_demangle_stacks` (if you don't already have an emscripten environment set up and don't want to bother just for this, I can run those for you).

**Comment by kripken** [comment]
The error on travis ci here looks like it's due to the binary size changing in a test. Can run `./auto_update_tests.py` to fix that (or just edit the output file manually).

**Comment by pipcet** [comment]
Thanks! Yeah, I verified the size change matches what should happen (again, if I'm reading the spec correctly): plus 5 bytes for the subsection size, 1 byte for the subsection type, 2x1 byte for the indices, minus 2 bytes for the missing local count. 

There's one slight change in behavior: the old code used to die when presented with a module that defined local names, the new code simply ignores the local[2] subsection of the "name" section. I think the new behavior is better, as I can run binaryen utilities directly on my wasm modules without having to strip the "name" section, but if you prefer the old behavior I can change it.

**Comment by kripken** [comment]
Yeah, that change makes sense. Although perhaps we should print a warning instead of silently ignoring it?

File: src/wasm/wasm-binary.cpp [hunk 1]
```diff
@@ -1511,23 +1524,27 @@ void WasmBinaryBuilder::readTableElements() {
 
 void WasmBinaryBuilder::readNames() {
   if (debug) std::cerr << "== readNames" << std::endl;
+  auto name_type = getU32LEB();
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp]: convention is `nameType` etc. for these new vars

File: src/wasm/wasm-binary.cpp [hunk 2]
```diff
@@ -1511,23 +1524,27 @@ void WasmBinaryBuilder::readTableElements() {
...
+  auto subsection_size = getU32LEB();
+  WASM_UNUSED(subsection_size);
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp]: we could use it at the end, to compare `pos` and see it incremented by the right amount? unless I'm missing something

**Comment by kripken** [comment]
Ok, just some minor comments, overall this looks correct, thanks @pipcet  :)

@lukewagner: you asked for a ping when we are ready to test this with a VM, this PR should be ready for that.

File: src/wasm/wasm-binary.cpp (hunk 2 above)
Comment (pipcet) [inline src/wasm/wasm-binary.cpp]: absolutely! And that actually caught a silly bug in my non-binaryen code, so thanks for the suggestion!

Comment (kripken) [inline src/wasm/wasm-binary.cpp]: Oh, heh, nice :)

Btw, is the other code you mention your gcc wasm backend? Are you connecting those in some way, or just using binaryen to inspect the output from gcc?

**Comment by lukewagner** [comment]
Hey thanks a lot!  So I don't know if yous have been following [design/#989](https://github.com/WebAssembly/design/issues/989), but @titzer's proposal is to name the updated extensible-names section `utf8-names` instead of just `names`.  Still need a PR on design repo to nail that down, though.  What's nice is this could allow engines to accept both `names` and `utf8-names` for a period of time, smoothing the transition.

**Comment by pipcet** [comment]
@lukewagner I confess I haven't been following that discussion, but I think that's an excellent idea (I'm not so sure utilities should support the `"name"` subsection still, but that's a separate issue from changing the name).

@kripken I suggest delaying this PR here until the design repo PR is through.

File: src/wasm/wasm-binary.cpp (hunk 2 above)
Comment (pipcet) [inline src/wasm/wasm-binary.cpp]: Just the latter, currently, but I'm looking at combining them when it makes sense to do so (I'm currently emitting `tee_local` at way too late a stage, by looking at the opcode stream for the right combinations, for example).

Comment (kripken) [inline src/wasm/wasm-binary.cpp]: I see, thanks. I'd be curious to see how the binaryen optimizer works on your output. It's a goal for it to be as general purpose as possible, so testing on more compilers can help.

**Comment by pipcet** [comment]
If I'm reading https://github.com/WebAssembly/design/pull/1016 correctly, the proposal to rename the `"name"` section has been dropped.

The question now is whether it's too late to handle only the new extensible name section, or whether some sort of autodetection (not very hard) is required to tell whether we're looking at an old-style or new-style `"name"` section. (Unless I'm mistaken, an old-style `"name"` section will never split into valid subsections).

In either case, I think it's time to make the extensible name section the default.

**Comment by dschuff** [comment]
I think we should just support the extensible name section. Speaking of which, do we have any tests that we drop bogus name sections? we should.

**Comment by dschuff** [comment]
V8 just landed a patch to update to extensible names section and I tested this patch against it. All of the aforementioned tests are passing, so I think this change can go in whenever we are ready.
As for when we are ready: basically, those tests break when binaryen doesn't match the JS engine. Since V8 is already updated and SM is presumably not, the right time to land it is presumably any time between now and when SM updates what it accepts. (at which point our waterfall with V8 will go green and those tests will cover this change, and the waterfall with SM will go red until it updates). Since it looks like the main emscripten buildbot is currently broken (and therefore we don't have SM test coverage right now anyway), I'd vote for just merging this without waiting.

**Comment by pipcet** [comment]
Thanks for testing!

I think the plan was for SM to update when binaryen does, so obviously I agree we should merge it soon.

**Comment by kripken** [comment]
I agree with @dschuff, we can merge this asap, unless @lukewagner has concerns?

**Comment by dschuff** [comment]
OK, our bots are able to cover this again, so hearing no objections I'm going to go ahead and merge it.

**Pull request merged**